ui/core-react/src/core-react/utils/hooks/useDisposable.ts

```diff
@@ -13,11 +13,31 @@ import type { IDisposable } from "@itwin/core-bentley";
  * A custom hook which creates a disposable object and manages its disposal on unmount
  * or factory method change.
  * @public
+ * @deprecated in 4.9.0. This hook is not compatible with React 18 Strict mode. Use [[useOptionalDisposable]] or `useState` + `useEffect` for
+ * creating and disposing disposable resources.
  */
 export function useDisposable<TDisposable extends IDisposable>(
   createDisposable: () => TDisposable
 ): TDisposable {
-  return useOptionalDisposable(createDisposable)!;
+  const [value, setValue] = useState(createDisposable());
```

Comment:
Why cant this keep using the new/fixed `useOptionalDisposable` ? If strict mode is running twice, why in this case are we trying to keep only one of the run instead of simply disposing of the first run and using the second, as we do in the optional one ?

Reply:
It cannot use new `useOptionalDisposable` implementation because it would break API. `useDisposable` function definition states that it never returns `undefined` but `useOptionalDisposable` is always returning `undefined` on first render.

The problem with `useDisposable` hook is that it has side effect, it creates disposable when initializing state and if `useEffect` if never run that disposable is never disposed. Changes made to StrictMode in React 18 helps to find this side effect because components are rendered twice. As a result `useDisposable` in StrictMode always leaks one disposable resource:
1. Component is rendered and `useDisposable` is called, disposable resource is created.
1. That component is thrown away. `useEffect` was never run, resource is not disposed.
1. Component is rendered again and `useDisposable` is called, disposabled resource is created.
1.  `useEffect` is run. New resource is not created due to `useInitialValue` flag.
1.  `useEffect` is run again. Cleanup function disposes `prevValue` (value that was created initializing state). Effect callback is run, new resource is created.

New `useDisposable` implementation does not try to keep only one of the runs. It tries to make sure that disposable resource created during render loop initializing state is disposed. I think it could be refactored a little bit to make it easier to read:
```tsx
const [value, setValue] = useState(createDisposable());
const initialValue = useRef(value);
useEffect(() => {
  if (initialValue.current) {
    initialValue.current.dispos();
    initialValue.current = undefined;
  }
  const newValue = createDisposable();
  setValue(newValue);
  return () => {
   newValue.dispose();
  };
}, [createDisposable]);
```

Reply:
Yes this part was missing, nice catch. 😞